### PR TITLE
notifications: Create stubs that save data to the database

### PR DESCRIPTION
**Pull request by DolceTriade**

### CODE REVIEW

**Comment by DolceTriade** [comment]
PTAL. updated with results of our discussion yesterday. App notifications still TODO.

File: backend/api_calls.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import logging
 import random
 import time
+import textwrap
```
Comment (IsaacG) [inline backend/api_calls.py:5]: I think this will requite updating the requirements.txt file

File: backend/api_calls.py [hunk 2]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+  """Queue a notification to be sent.
+
+  Validation:
+    notificationId does not exist.
+    groupId or playerId exists.
+
+  Args:
+    notificationId: Queued Notification id to create.
+    message: Message to send to client.
+    previewMessage: Short preview of the message conents. Optional.
```
Comment (IsaacG) [inline backend/api_calls.py]: All the other non-update methods don't have any optional arg. Could we require this arg be present even if not set?

File: backend/api_calls.py [hunk 3]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+    app: Whether to send as a push notification to the iOS/Android/Web clients.
+    vibrate: Whether the notification should vibrate on iOS/Android.
+    sound: Whether the notification should play a sound on iOS/Android.
+    destination: URL that notification should open when clicked.
+    sendTime: When to spawn notifications from this template. UNIX timestamp.
+    playerId: Player id to send message to.
+    groupId: Who to send the notifications to.
+    icon: An icon code to show.
+
```
Comment (IsaacG) [inline backend/api_calls.py:899]: Note: It's groupId or playerId but not both.

File: backend/api_calls.py [hunk 4]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+  Firebase entries:
+    /notifications/%(notificationId)
+"""
+  valid_args = ['!notificationId']
+  if 'groupId' in request:
+    valid_args.append('groupId')
+  elif 'playerId' in request:
+    valid_args.append('playerId')
+  else:
+    raise InvalidInputError('Must include either a playerId or a groupId')
```
Comment (IsaacG) [inline backend/api_calls.py]: The valid_args only validate if present so you can just dump both in.

File: backend/api_calls.py [hunk 5]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+
+  if 'previewMessage' not in request:
+    request['previewMessage'] = textwrap.wrap(request['message'], 100)[0]
+
+  put_data = {}
+  properties = ['message', 'app', 'vibrate', 'sound', 'destination', 'sendTime',
+                'groupId', 'playerId', 'icon', 'previewMessage']
+
+  for property in properties:
+    if property in request:
```
Comment (IsaacG) [inline backend/api_calls.py:927]: If they are required, there's no need to test for presence

File: backend/api_calls.py [hunk 6]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+    /notifications/%(notificationId)
+"""
+  valid_args = ['notificationId']
+  required_args = list(valid_args)
+  ValidateInputs(request, firebase, required_args, valid_args)
+
+  put_data = {}
+  properties = ['message', 'app', 'vibrate', 'sound', 'destination', 'sendTime',
+                'groupId', 'playerId', 'icon']
+
```
Comment (IsaacG) [inline backend/api_calls.py:963]: Validate the old sendTime and (if present) new sendTime didn't pass yet.

File: backend/api_calls.py [hunk 7]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+  """Updates the notification's seenTime.
+
+  Validation:
+    notificationId must exist.
+
+  Args:
+    notificationId: Notification Id to update.
+
+  Firebase entries:
+    /notifications/notifications/%(notificationId)
```
Comment (IsaacG) [inline backend/api_calls.py]: This needs to be done on a per-player basis. I think you want /players/%(id)/notifications ?

File: backend/api_calls.py [hunk 8]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+  put_data = {
+    'seenTime': time.time()
+  }
+  results = []
+  results.append(firebase.patch('/notifications/notifications/%s' % (
+      request['notificationId']), put_data))
+  return results
+
+
+def _handleNotification(self, firebase, notification_id, notification):
```
Comment (IsaacG) [inline backend/api_calls.py]: 1. I'm not a fan of changing the naming convention to signal helper methods. If you don't have a strong opinion, let's keep it consistent.
2. There's no "self" outside classes.

File: backend/api_calls.py [hunk 9]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+  """Helper function to propogate a notification."""
+
+  # TODO: Send to ionic
+  if 'playerId' in notification:
+    firebase.put('/players/%s/notifications' % notification['playerId'],
+                  notification_id, notification)
+  elif 'groupId' in notification:
+    players = firebase.get('/groups/%s' % notification['groupId'],
+                           'players')
+    logging.info(players)
```
Comment (IsaacG) [inline backend/api_calls.py]: If you're leaving this line in here, make it more descriptive.

File: backend/api_calls.py [hunk 10]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+    for player in players:
+      firebase.put('/players/%s/notifications' % player,
+                   notification_id, notification)
+  else:
+    logging.error('Notification does not have a playerId or a groupId!')
```
Comment (IsaacG) [inline backend/api_calls.py]: Including the notification_id here would be helpful

File: backend/api_calls.py [hunk 11]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+  # notification database.
+  get_params = {
+      #
+      'orderBy': '"sent"',
+      'limitToLast': 20
+  }
+  updates = True
+  while updates:
+    updates = False
+    notifications = firebase.get('/', 'notifications', params=get_params)
```
Comment (IsaacG) [inline backend/api_calls.py]: You can also do get('/notifications', None, params)

File: backend/api_calls.py [hunk 12]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+def ExecuteNotifications(request, firebase):
+  """INTERNAL ONLY: Send the notifications.
+
+  This method will execute queued notifications and send them to apps and add
+  them to the player specific notification database URLs.
+  """
+  # Handle notifications, 20 at a time so we don't need to get the entire
+  # notification database.
+  get_params = {
+      #
```
Comment (IsaacG) [inline backend/api_calls.py]: Empty comment?

File: backend/api_calls.py [hunk 13]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+      'orderBy': '"sent"',
+      'limitToLast': 20
+  }
+  updates = True
+  while updates:
+    updates = False
+    notifications = firebase.get('/', 'notifications', params=get_params)
+    if not notifications:
+      return
+    logging.info(notifications)
```
Comment (IsaacG) [inline backend/api_calls.py]: If you leave this in, add context to the log line

File: backend/api_calls.py [hunk 14]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+    current_time = time.time()
```
Comment (IsaacG) [inline backend/api_calls.py]: This should probably be done once prior to the loop

File: backend/api_calls.py [hunk 15]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+    for notification_id, notification in notifications.iteritems():
+      if 'sent' not in notification and current_time > notification['sendTime']:
+        updates = True
+        _handleNotification(firebase, notification_id, notification)
+        notifications[notification_id]['sent'] = True
+    if updates:
+      result = firebase.patch('/notifications', notifications)
+      if not result:
```
Comment (IsaacG) [inline backend/api_calls.py]: if result is None:

File: backend/api_calls.py [hunk 16]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+        logging.error('Error updating notification results...')
```
Comment (IsaacG) [inline backend/api_calls.py]: Include the ID and what sort of update this is.

File: backend/api_calls.py [hunk 17]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+    notificationId: Notification Id to update.
+
+  Firebase entries:
+    /notifications/notifications/%(notificationId)
+"""
+  valid_args = ['notificationId', 'playerId']
+  required_args = list(valid_args)
+  ValidateInputs(request, firebase, required_args, valid_args)
+  put_data = {
+    'seenTime': time.time()
```
Comment (IsaacG) [inline backend/api_calls.py]: To reduce the number of unique key names, please use just 'time'

File: backend/api_calls.py [hunk 18]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+  }
+  results = []
+  results.append(firebase.patch('/notifications/notifications/%s' % (
```
Comment (IsaacG) [inline backend/api_calls.py]: For a single Firebase call, just return the result; there's no value in wrapping it in an array

File: backend/api_calls.py [hunk 19]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+      request['notificationId']), put_data))
+  return results
+
+
+def _handleNotification(self, firebase, notification_id, notification):
+  """Helper function to propogate a notification."""
+
+  # TODO: Send to ionic
+  if 'playerId' in notification:
```
Comment (IsaacG) [inline backend/api_calls.py]: You can do:   
if playerId: players = [playerId]
elif groupId: players = firebase.get(...)

Then you don't have to have the firebase.put line duplicated.

File: backend/api_calls.py [hunk 20]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+                           'players')
+    logging.info(players)
+    for player in players:
+      firebase.put('/players/%s/notifications' % player,
+                   notification_id, notification)
+  else:
+    logging.error('Notification does not have a playerId or a groupId!')
+
+
+def ExecuteNotifications(request, firebase):
```
Comment (IsaacG) [inline backend/api_calls.py]: This isn't part of the /api methods so should probably go in its own file.
Splitting this into its own PR wouldn't hurt to keep commits reasonably sized and allow a commit message to capture them.

File: backend/api_calls.py [hunk 21]
```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
...
+  """
+  # Handle notifications, 20 at a time so we don't need to get the entire
+  # notification database.
+  get_params = {
+      #
+      'orderBy': '"sent"',
+      'limitToLast': 20
+  }
+  updates = True
+  while updates:
```
Comment (IsaacG) [inline backend/api_calls.py]: If Firebase sends sent notifications only in the first batch, does this loop exit before trying to see if there are other notifications?

File: backend/main.py [hunk 22]
```diff
@@ -86,6 +84,9 @@ def HandleError(e):
   'addReward': api_calls.AddReward,
   'addRewards': api_calls.AddRewards,
   'claimReward': api_calls.ClaimReward,
+  'sendNotification': api_calls.SendNotification,
+  'updateNotification': api_calls.UpdateNotification,
+  'markNotificationSeen': api_calls.MarkNotificationSeen,
```
Comment (IsaacG) [inline backend/main.py:90]: It'd be nice to see these in the backend end2end test

File: backend/main.py [hunk 23]
```diff
@@ -114,6 +115,15 @@ def GetGun():
   return jsonify(GetFirebase().get('/guns', gun))
 
 
+@app.route('/cronNotification', methods=['GET'])
+def CronNotification():
+  CRON_KEY = 'X-Appengine-Cron'
```
Comment (IsaacG) [inline backend/main.py]: This isn't set up as a global constant so it shouldn't be all-caps

File: backend/api_calls.py (hunk 1 above)
Comment (DolceTriade) [reply backend/api_calls.py:5]: I believe this is a default python module.

File: backend/api_calls.py (hunk 2 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Done.

File: backend/api_calls.py (hunk 4 above)
Comment (DolceTriade) [reply backend/api_calls.py]: But I want to make sure they don't include them both (though, I can accept both, but give one precedence. It doesn't really matter to me.)

File: backend/api_calls.py (hunk 3 above)
Comment (DolceTriade) [reply backend/api_calls.py:899]: Done.

File: backend/api_calls.py (hunk 5 above)
Comment (DolceTriade) [reply backend/api_calls.py:927]: groupId and playerId are mutually exclusive.

File: backend/api_calls.py (hunk 6 above)
Comment (DolceTriade) [reply backend/api_calls.py:963]: Done.

File: backend/api_calls.py (hunk 1 above)
Comment (IsaacG) [reply backend/api_calls.py:5]: My bad :)

File: backend/api_calls.py (hunk 7 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Whoops, you're right.

File: backend/api_calls.py (hunk 4 above)
Comment (IsaacG) [reply backend/api_calls.py]: I would put both into the valid_args to keep that simple then be explicit about that test.
if 'groupId' in request and 'playerId' in request:
  raise InvalidInput('You cannot do both')

File: backend/api_calls.py (hunk 5 above)
Comment (IsaacG) [reply backend/api_calls.py:927]: Oh, yeah. I suppose splitting those two out from the rest of the args would be even more messy.

File: backend/api_calls.py (hunk 17 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Done

File: backend/api_calls.py (hunk 18 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Done.

File: backend/api_calls.py (hunk 8 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Whoops, copied out of a notifications class I made before I learned about Google App Engine Cron. Cleaned up lingering self reference.

File: backend/api_calls.py (hunk 19 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Good idea.

File: backend/api_calls.py (hunk 10 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Done.

File: backend/api_calls.py (hunk 9 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Removed.

File: backend/main.py (hunk 23 above)
Comment (IsaacG) [reply backend/main.py]: This built in GAE cron with built in auth stuff is pretty nifty. Nice find!

File: backend/api_calls.py (hunk 11 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Done.

File: backend/api_calls.py (hunk 21 above)
Comment (DolceTriade) [reply backend/api_calls.py]: It keeps checking until it doesn't send any more notifications. Modified code a bit to make this more clear.

File: backend/api_calls.py (hunk 14 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Done.

File: backend/api_calls.py (hunk 15 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Done.

File: backend/api_calls.py (hunk 16 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Done.

File: backend/main.py (hunk 23 above)
Comment (DolceTriade) [reply backend/main.py]: Done.

File: backend/main.py (hunk 22 above)
Comment (DolceTriade) [reply backend/main.py:90]: Will add tests next.

**Comment by DolceTriade** [comment]
Thanks for the review, uploaded changes.

File: backend/api_calls.py [hunk 24]
```diff
@@ -877,6 +877,132 @@ def RandomWords(n):
...
+  properties = ['message', 'app', 'vibrate', 'sound', 'destination', 'sendTime',
+                'groupId', 'playerId', 'icon']
+
+  current_time = time.time()
+  if 'sendTime' in request and current_time > int(request['sendTime']):
+    raise InvalidInputError('sendTime must not be in the past!')
+
+  notification = firebase.get('/notifications', request['notificationId'])
+  # This shouldn't happen since we validated this above...
+  if not notification:
```
Comment (IsaacG) [inline backend/api_calls.py:970]: Note this is already tested by the Validate()

File: backend/api_calls.py [hunk 25]
```diff
@@ -877,6 +877,132 @@ def RandomWords(n):
...
+    playerId: Player who saw notification.
+
+  Firebase entries:
+    /players/%(playerId)/notifications/%(notificationId)
+"""
+  valid_args = ['notificationId', 'playerId']
+  required_args = list(valid_args)
+  ValidateInputs(request, firebase, required_args, valid_args)
+  put_data = {
+    'time': time.time()
```
Comment (IsaacG) [inline backend/api_calls.py:1000]: Note: time.time() returns a float, not a UNIX timestamp. Not sure if we should stick to timestamps or not.
For timestamp, int(time.time())

File: backend/main.py [hunk 26]
```diff
@@ -114,6 +116,15 @@ def GetGun():
   return jsonify(GetFirebase().get('/guns', gun))
 
 
+@app.route('/cronNotification', methods=['GET'])
+def CronNotification():
+  cron_key = 'X-Appengine-Cron'
+  if cron_key not in request.headers or not request.headers[cron_key]:
+    return 'Unauthorized', 403
+  notifications.ExecuteNotifications(None, GetFirebase())
```
Comment (IsaacG) [inline backend/main.py:124]: Does it make sense to return the result ot ExecuteNotification() ?

File: backend/notifications.py [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+def ExecuteNotifications(request, firebase):
+  """INTERNAL ONLY: Send the notifications.
+
+  This method will execute queued notifications and send them to apps and add
+  them to the player specific notification database URLs.
+  """
+  # Handle notifications, 20 at a time so we don't need to get the entire
+  # notification database.
+  get_params = {
+      'orderBy': '"sent"',
```
Comment (IsaacG) [inline backend/notifications.py:33]: I can't find any documentation for this params stuff. Can you point me to it? Are you certain the orderBy is expecting a string containing quotes?

File: backend/notifications.py [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+      'limitToLast': 20
+  }
+  current_time = time.time()
+  while True:
+    updates = False
+    notifications = firebase.get('/notifications', None, params=get_params)
+    if not notifications:
```
Comment (IsaacG) [inline backend/notifications.py]: This should probably also be:   if notifications is None

File: backend/notifications.py [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+      return
+    logging.info(notifications)
```
Comment (IsaacG) [inline backend/notifications.py]: This output could use some context in the logs or drop it if it's for dev debugging only.

**Review by IsaacG — APPROVED** [review]
Looks good.  I got some comments that should be addressed before merging but once addressed, feel free to merge or ping me to merge it.

File: backend/main.py (hunk 26 above)
Comment (DolceTriade) [reply backend/main.py:124]: I thought about it, but I don't think cron actually cares, so I only return something if someone other than cron tries to hit that endpoint.

File: backend/notifications.py (hunk 27 above)
Comment (DolceTriade) [reply backend/notifications.py:33]: Yes, it needs to be a valid json string. https://firebase.google.com/docs/database/rest/retrieve-data#section-complex-queries

File: backend/notifications.py (hunk 29 above)
Comment (DolceTriade) [reply backend/notifications.py]: blah, more debug logs... removed.

File: backend/notifications.py (hunk 28 above)
Comment (DolceTriade) [reply backend/notifications.py]: Done.

File: backend/api_calls.py (hunk 25 above)
Comment (DolceTriade) [reply backend/api_calls.py:1000]: Done.

File: backend/notifications.py (hunk 27 above)
Comment (IsaacG) [reply backend/notifications.py:33]: Oh, cool. There's a lot of useful stuff there I ought to use, esp things like shallow=true. Thanks!

**Pull request merged**